lib/Db/FailedLinkAccessMapper.php

```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * @author Semih Serhat Karakaya <karakayasemi@itu.edu.tr>
+ *
+ * @copyright Copyright (c) 2019 ownCloud GmbH
+ * @license GPL-2.0
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+namespace OCA\BruteForceProtection\Db;
+
+use OCP\AppFramework\Db\Mapper;
+use OCA\BruteForceProtection\BruteForceProtectionConfig;
+use OCP\AppFramework\Utility\ITimeFactory;
+use OCP\IDBConnection;
+
+/**
+ * Class FailedLinkAccessMapper
+ * @package OCA\BruteForceProtection\Db
+ */
+class FailedLinkAccessMapper extends Mapper {
+
+	/**
+	 * @var BruteForceProtectionConfig $config
+	 */
+	protected $config;
+
+	/**
+	 * @var ITimeFactory $timeFactory
+	 */
+	protected $timeFactory;
+
+	/**
+	 * @var string $tableName
+	 */
+	protected $tableName = 'bfp_link_accesses';
+
+	/**
+	 * FailedLinkAccessMapper constructor.
+	 *
+	 * @param IDBConnection $db
+	 * @param BruteForceProtectionConfig $config
+	 * @param ITimeFactory $timeFactory
+	 */
+	public function __construct(
+		IDBConnection $db,
+		BruteForceProtectionConfig $config,
+		ITimeFactory $timeFactory
+	) {
+		parent::__construct($db, $this->tableName);
+		$this->config = $config;
+		$this->timeFactory = $timeFactory;
+	}
+
+	/**
+	 * @param string $token
+	 * @param string $ip
+	 * @return int
+	 */
+	public function getFailedAccessCountForTokenIpCombination($token, $ip) {
+		$builder = $this->db->getQueryBuilder();
+		$thresholdTime = $this->getLastFailedAccessTimeForTokenIpCombination($token, $ip) - $this->config->getBruteForceProtectionTimeThreshold();
```

Comment:
Any better alternative? I'd rather not call to public methods (of the same class) because code could be reported as covered by tests although they could be missing; in addition, there could be concurrency problems affecting the accuracy of the result.
Just double-checking that at least this won't be critical

Reply:
now method takes `$thresholdTime` as a parameter and does not call its own method anymore.